tests/DCPS/MultiDPTest/publisher.cpp

```diff
@@ -160,165 +127,131 @@ void init ()
-
-      init ();
-
-      // Indicate that the publisher is ready
-      FILE* writers_ready = ACE_OS::fopen (pub_ready_filename.c_str (), ACE_TEXT("w"));
-      if (writers_ready == 0)
-        {
-          ACE_ERROR ((LM_ERROR,
-                      ACE_TEXT("(%P|%t) ERROR: Unable to create publisher ready file\n")));
-        }
-
-      const ACE_Time_Value small_time(0,250000);
-      // Wait for the subscriber to be ready.
-      FILE* readers_ready = 0;
-      do
-        {
-          ACE_OS::sleep (small_time);
-          readers_ready = ACE_OS::fopen (sub_ready_filename.c_str (), ACE_TEXT("r"));
-        } while (0 == readers_ready);
-
-      if (writers_ready) ACE_OS::fclose(writers_ready);
-      if (readers_ready) ACE_OS::fclose(readers_ready);
-
-      // ensure the associations are fully established before writing.
-      ACE_OS::sleep(3);
-
-      {  // Extra scope for VC6
-        for (int i = 0; i < num_datawriters; i ++)
-          {
-            writers[i]->start ();
-          }
-      }
+  try {
+    dpf = TheParticipantFactoryWithArgs(argc, argv);
+    ACE_DEBUG((LM_INFO, "(%P|%t) %T publisher main\n"));
 
-      int timeout_writes = 0;
-      bool writers_finished = false;
-
-      while ( !writers_finished )
-        {
-          writers_finished = true;
-          for (int i = 0; i < num_datawriters; i ++)
-            {
-              writers_finished = writers_finished && writers[i]->is_finished();
-            }
-          ACE_OS::sleep(small_time);
-        }
-
-      {  // Extra scope for VC6
-        for (int i = 0; i < num_datawriters; i ++)
-          {
-            timeout_writes += writers[i]->get_timeout_writes();
-          }
-      }
-      // Indicate that the publisher is done
-      FILE* writers_completed = ACE_OS::fopen (pub_finished_filename.c_str (), ACE_TEXT("w"));
-      if (writers_completed == 0)
-        {
-          ACE_ERROR ((LM_ERROR,
-                      ACE_TEXT("(%P|%t) ERROR: Unable to i publisher completed file\n")));
-        }
-      else
-        {
-          ACE_OS::fprintf (writers_completed, "%d\n", timeout_writes);
-        }
-
-      // Wait for the subscriber to finish.
-      FILE* readers_completed = 0;
-      do
-        {
-          ACE_OS::sleep (small_time);
-          readers_completed = ACE_OS::fopen (sub_finished_filename.c_str (), ACE_TEXT("r"));
-        } while (0 == readers_completed);
-
-      if (writers_completed) ACE_OS::fclose(writers_completed);
-      if (readers_completed) ACE_OS::fclose(readers_completed);
-
-      {  // Extra scope for VC6
-        for (int i = 0; i < num_datawriters; i ++)
-          {
-            writers[i]->end ();
-            delete writers[i];
-          }
+    // let the Service_Participant (in above line) strip out -DCPSxxx parameters
+    // and then get application specific parameters.
+    parse_args(argc, argv);
+
+    init();
+
+    Allocator* allocator;
+    ACE_NEW_RETURN(allocator, Allocator(mmap_file), -1);
```

Comment:
Is this going to work on all supported platforms?  It's unclear from the diffs on GitHub where this class is coming from.

The test is set up to use temporary files since they are more widely supported.  They can also support running tests across different hosts - this has been used with embedded targets where one side (perhaps publisher) is on the target and the other (perhaps subscriber) is on the host system.  Using a shared filesystem like NFS allows the temp files to work across hosts.

Reply:
`Allocator` is a typedef of an `ACE_Malloc_T` template class. My intention was to replace the temporary files with shared memory for synchronization between publisher and subscriber. But this wouldn't work if the publisher and subscriber run on different hosts. 

The reason I changed to use shared memory is because one of the temporary files was also used to communicate the number of timed-out writes. But it's possible that the subscriber may fail to read that number even though the publisher has written that number (but it isn't flushed to the file yet). But I think that could be fixed to keep using temporary files.